### PR TITLE
fix: Fix server side skip test cases

### PR DESCRIPTION
**Pull request by sagar-qa007**
Fixes #`Issue Number`  
_or_  
Fixes `Issue URL`
> [!WARNING]  
> _If no issue exists, please create an issue first, and check with the maintainers if the issue is valid._

## Automation

/ok-to-test tags="@tag.IDE"

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11370534880>
> Commit: dea75ff1999643d072745a38e91f741677ac5703
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11370534880&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.IDE`
> Spec:
> <hr>Wed, 16 Oct 2024 17:40:56 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved handling and reporting of test failures during the build process.
	- Enhanced logging for better clarity on test results.

- **New Features**
	- Introduced detailed output for failed and skipped tests, now available in the GitHub step summary and as comments on pull requests.
	- Added new input parameter `is-pg-build` for enhanced workflow control.
	- Updated default value for `skip-tests` to "false" to enforce test execution.

### CODE REVIEW

**Comment by sagar-qa007** [comment]
@abhvsn can we please review this ?

**Review by abhvsn — CHANGES_REQUESTED** [review]
Can we keep the failed and skipped tests separately may be within the same file so that if needed to skip the test deliberately it should not get picked up on next re-run?

File: .github/workflows/server-build.yml [hunk 1]
```diff
@@ -176,39 +176,94 @@ jobs:
...
-          if ! mvn test "${args[@]}"; then
-            echo "Generating failed test list:"
-            failed_tests_file="$PWD/failed-server-tests.txt"
-            gawk -F\" '/<testcase / {cur_test = $4 "#" $2} /<(failure|error) / {print cur_test}' $(find . -type f -name 'TEST-*.xml') \
-              | sort -u \
-              | tee "$failed_tests_file"
-            echo "Saved to $failed_tests_file"
-            if [[ -s $failed_tests_file ]]; then
+
+          echo "test_result variable value: ${test_result}"
```
[the remark this answers is not in the document]
Comment (sagar-qa007) [reply .github/workflows/server-build.yml:199]: For re-run it should be in a file. If need to keep separate then we need to update the whole flow.

File: .github/workflows/server-build.yml [hunk 2]
```diff
@@ -176,39 +176,94 @@ jobs:
...
+              module_directories=$(find . -path "*/${module}*/src/test/java/*" -type f -name "*Test.java" -exec dirname {} \; | sort -u)
+              for module_directory in $module_directories; do
+                  test_classes=$(find "$module_directory" -type f -name "*Test.java" | sed 's|.*/src/test/java/||; s|\.java$||; s|/|.|g')
+                  for class_name in $test_classes; do
+                      if [[ ${#class_name} -le 240 ]] && ! grep -Fxq "$class_name#" "$OUTPUT_FILE"; then
+                          echo "${class_name}#" >> "$OUTPUT_FILE"
+                      fi
+                  done
+              done
+          done
```
[the remark this answers is not in the document]
Comment (sagar-qa007) [reply .github/workflows/server-build.yml:242]: Currently we add all the cases from the failed module. I can improve on this by checking the xml file. I am considering that we need to re-running the whole module again if case failed.

**Comment by sagar-qa007** [comment]
@abhvsn Please check if we can approve this.

**Review by abhvsn — COMMENTED** [review]
@sagar-qa007 IIUC we are running all the tests in a module in case of partial failure or skipped tests. Considering the recent flakiness in appsmith-server module can't we just run the failed tests? This will increase the reruns and the overall time to pass the server tests.

**Review by abhvsn — APPROVED** [review]
LGTM 

I hope we have checked this against the actual run with skip and failed tests.

**Pull request merged**